### PR TITLE
Fix some clippy lints for library/std/src/sys/windows

### PR DESCRIPTION
**Pull request by ChrisDenton**
These issues were shown by running `x clippy` on `library/std` and filtering for `windows/` paths. I think running clippy on the full std would be great but I wanted to start smaller and with something that's hopefully easier to review. It'd be good to eventually run clippy in CI but that's a bigger conversation.

I've created separate commits for each clippy lint fixed (with the commit title set to the lint name) and reviewed the changes myself. Most of the fixes here are trivial.

r? libs

### CODE REVIEW

**Comment by ChrisDenton** [comment]
I was so busy playing with clippy that I forgot fmt.

**Comment by matthiaskrgr** [comment]
should definitely get a perf run when you think its ready.

**Comment by ChrisDenton** [comment]
A perf run is very unlikely to show any changes seeing as this is Windows only. Also I'm not really convinced any of these lints would affect perf. I think the only ones that *might* potentially (and I'm sceptical) are `unnecessary_lazy_evaluations`, `manual_slice_size_calculation`, `duration_subsec` and `manual_range_contains`. Of those, only `manual_range_contains` is in perf sensitive code (utf-8 <--> utf-16 conversion).

File: library/std/src/sys/windows/stdio.rs [hunk 1]
```diff
@@ -207,7 +207,7 @@ fn write_valid_utf8_to_console(handle: c::HANDLE, utf8: &str) -> io::Result<usiz
         // write the missing surrogate out now.
         // Buffering it would mean we have to lie about the number of bytes written.
         let first_code_unit_remaining = utf16[written];
-        if first_code_unit_remaining >= 0xDCEE && first_code_unit_remaining <= 0xDFFF {
+        if matches!(first_code_unit_remaining, 0xDCEE..=0xDFFF) {
```
Comment (ChrisDenton) [inline library/std/src/sys/windows/stdio.rs:210]: I confirmed that using `matches!` is exactly equivalent to using a manual if (in terms of codegen) and I do think it's clearer.

File: library/std/src/sys/windows/io.rs [hunk 2]
```diff
@@ -36,7 +36,7 @@ impl<'a> IoSlice<'a> {
 
     #[inline]
     pub fn as_slice(&self) -> &[u8] {
-        unsafe { slice::from_raw_parts(self.vec.buf as *mut u8, self.vec.len as usize) }
```
[the remark this answers is not in the document]
Comment (ChrisDenton) [reply library/std/src/sys/windows/io.rs:39]: That's true. Though we're a bit at the mercy of the [win32metadata project](https://github.com/microsoft/win32metadata) here (which is what our bindings are ultimately generated from) which might not necessarily follow what we think a `c_char` is (as it has no particular need to follow C).

File: library/std/src/sys/windows/fs.rs [hunk 3]
```diff
@@ -156,7 +156,7 @@ impl DirEntry {
     }
 
     pub fn path(&self) -> PathBuf {
-        self.root.join(&self.file_name())
+        self.root.join(self.file_name())
```
[the remark this answers is not in the document]
Comment (ChrisDenton) [reply library/std/src/sys/windows/fs.rs:159]: That's `needless_borrows_for_generic_args`. There's also the related `needless_borrow` which found a lot too!

**Pull request merged**